tests/unit/s2n_client_signature_algorithms_extension_test.c

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+
+#include <stdint.h>
+
+#include "tls/s2n_config.h"
+#include "tls/s2n_connection.h"
+#include "tls/s2n_client_extensions.h"
+#include "tls/s2n_tls.h"
+
+#include "stuffer/s2n_stuffer.h"
+#include "utils/s2n_safety.h"
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    {
+        /* Test that unknown TLS_EXTENSION_SIGNATURE_ALGORITHMS values are ignored and negotiation fails */
+        struct s2n_sig_scheme_list sig_hash_algs = {
+            .iana_list = { 0xFF01, 0xFFFF },
+            .len = 2,
+        };
+        struct s2n_connection *conn;
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+
+        struct s2n_stuffer signature_algorithms_extension;
+        EXPECT_SUCCESS(s2n_stuffer_alloc(&signature_algorithms_extension, 2 + (sig_hash_algs.len * 2)));
+        GUARD(s2n_stuffer_write_uint16(&signature_algorithms_extension, sig_hash_algs.len * 2));
+        for (int i = 0; i < sig_hash_algs.len; i++) {
+            GUARD(s2n_stuffer_write_uint16(&signature_algorithms_extension, sig_hash_algs.iana_list[i]));
+        }
+
+        struct s2n_array *parsed_extensions = s2n_array_new(sizeof(struct s2n_client_hello_parsed_extension));
+        struct s2n_client_hello_parsed_extension *parsed_named_group_extension = s2n_array_pushback(parsed_extensions);
+        parsed_named_group_extension->extension_type = TLS_EXTENSION_SIGNATURE_ALGORITHMS;
+        parsed_named_group_extension->extension = signature_algorithms_extension.blob;
+
+        /* If only unknown algorithms are offered, expect choosing a scheme to fail */
+        EXPECT_SUCCESS(s2n_client_extensions_recv(conn, parsed_extensions));
+        EXPECT_EQUAL(conn->handshake_params.client_sig_hash_algs.len, sig_hash_algs.len);
+        EXPECT_FAILURE(s2n_choose_sig_scheme_from_peer_preference_list(conn, &conn->handshake_params.client_sig_hash_algs,
```

Comment:
Since these tests are really for s2n_choose_sig_scheme_from_peer_preference_list, should this file be named s2n_signature_algorithms_test? And do we need the s2n_client_extensions_recv part?

Reply:
The impetus for writing the test was to make sure extensions handle unknown values correctly. I wanted to make sure I setup the extensions as they would be received by a client, and then verify the expected behavior after processing the extension using `s2n_client_extensions_recv()`.

While the failure is expected from `s2n_choose_sig_scheme_from_peer_preference_list()`, the test is making sure the function fails as expected when the connection's sig_hash_algs has been setup by the extension processor.

As for the name of the test file, I'm not sure what is best. I can certainly rename it.